src/VisualStudio/Core/Def/Implementation/Diagnostics/DiagnosticProgressReporter.cs

```diff
@@ -22,6 +25,28 @@ internal sealed class TaskCenterSolutionAnalysisProgressReporter
         private TaskCompletionSource<VoidResult> _currentTask;
         private DateTimeOffset _lastTimeReported;
 
+        private int _lastPendingItemCount;
+        private ProgressStatus _lastProgressStatus;
+        private ProgressStatus _lastShownProgressStatus;
+
+        // _resettableDelay makes sure that we at the end update status to correct value.
+        // in contrast to _lastTimeReported makes sure that we only update at least s_minimumInterval interval.
+        //
+        // for example, when an event stream comes in as below (assuming 200ms minimum interval)
+        // e1 -> (100ms)-> e2 -> (300ms)-> e3 -> (100ms) -> e4
+        //
+        // actual status shown to users without _resettableDelay will be 
+        // e1 -> e3.
+        //
+        // e2 and e4 will be skipped since interval was smaller than min interval.
+        // losing e2 is fine, but e4 is problematic since the user now could see the wrong status 
+        // until the next event comes in. 
+        // for example, it could show "Evaluating" when it is actually "Paused" until the next event
+        // which could be long time later.
+        // what _resettableDelay does is making sure that if the next event doesn't come in 
+        // within certain delay, it updates status to e4 (current).
+        private ResettableDelay _resettableDelay;
```

Comment:
note: an extremely simple alternative (used elsewhere in roslyn) is simply to have a update task with a reasonable delay associated with it (maybe 250ms).  If you need to update the UI, you just see if that update task is in flight, if so you can bail.  You then only make the update delay if there's no in-flight task. This can be done in a threadsafe fashion with a simple lock used to just guard the reading of the task, or the reading+creation of the task. 

We have this in several ther places and it's both super simple, safe, and ensures a fixed point is reached.

Reply:
basically, it addresses the only real concern with UI updating.  Namely needing to throttle it so we don't swamp the UI thread.  We tested this fairly extensively and found it works great and doesn't have any negative impacts. (i.e. the lock taking is basically no perf cost at all).

Reply:
that makes UI to update only when there is no event coming in for a certain amount of time. this periodically updates UI regardless of how fast events are coming for how long. and I don't like the task continued with a lot of cancellation as well. ResettableDelay is a way cheaper approach.  I believe it was introduced long time ago when perf team discovered our approach of task chain with cancellation is unnecessarily heavy for simple job.
sure, there could be case where task chain with cancellation can have a benefit, but in this case, I dont see any.

Reply:
ah, misunderstood your approach.  that approach still doesn't remove the need for me to have resettable delay here. since you can miss the last event and there is no way to get it back.

that approach only works with pulling and as I said, pulling make status to almost always says "paused".

Reply:
>  since you can miss the last event and there is no way to get it back.

What do you mean by "miss the last event"? The idea here is that instead of trying to stream (or push) events to the UI, you simple ensure the UI updates at a reasonable pace and the UI *pulls* the data it needs when it wants to actually update.  

> random "Evalutiong" or "Paused" for up to the delay 

In practice, with a normal delay (i.e. 100-250ms) users wouldn't be able to perceive anything.  The UI simply updates rapidly with the correct information.  :) 

> I want "Evaluating" to show up even with the throttle as long as possible until we have a big "Paused" period.

Ok.  It was just a suggestion that i thought would simplify.  Up to you :)